File: android/src/main/kotlin/com/mapbox/maps/mapbox_maps/MapboxMapController.kt
package com.mapbox.maps.mapbox_maps

import android.content.Context
import android.view.View
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import com.google.gson.Gson
import com.mapbox.common.*
import com.mapbox.maps.*
import com.mapbox.maps.mapbox_maps.annotation.AnnotationController
import com.mapbox.maps.pigeons.FLTMapInterfaces
import com.mapbox.maps.pigeons.FLTSettings
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

class MapboxMapController(
  context: Context,
  mapInitOptions: MapInitOptions,
  private val lifecycleProvider: MapboxMapsPlugin.LifecycleProvider,
  eventTypes: List<Int>,
  messenger: BinaryMessenger,
  channelSuffix: Int,
  pluginVersion: String
) : PlatformView,
  DefaultLifecycleObserver,
  MethodChannel.MethodCallHandler {

  private val mapView: MapView = MapView(context, mapInitOptions)
  private val mapboxMap: MapboxMap = mapView.mapboxMap
  private val methodChannel: MethodChannel
  private val styleController: StyleController = StyleController(mapboxMap, context)
  private val cameraController: CameraController = CameraController(mapboxMap, context)
  private val projectionController: MapProjectionController = MapProjectionController(mapboxMap)
  private val mapInterfaceController: MapInterfaceController = MapInterfaceController(mapboxMap, context)
  private val animationController: AnimationController = AnimationController(mapboxMap, context)
  private val annotationController: AnnotationController = AnnotationController(mapView, mapboxMap)
  private val locationComponentController = LocationComponentController(mapView, context)
  private val gestureController = GestureController(mapView)
  private val logoController = LogoController(mapView)
  private val attributionController = AttributionController(mapView)
  private val scaleBarController = ScaleBarController(mapView)
  private val compassController = CompassController(mapView)

  private val proxyBinaryMessenger = ProxyBinaryMessenger(messenger, "/map_$channelSuffix")
  init {
    changeUserAgent(pluginVersion)
    lifecycleProvider.getLifecycle()?.addObserver(this)
    FLTMapInterfaces.StyleManager.setup(proxyBinaryMessenger, styleController)
    FLTMapInterfaces._CameraManager.setup(proxyBinaryMessenger, cameraController)
    FLTMapInterfaces.Projection.setup(proxyBinaryMessenger, projectionController)
    FLTMapInterfaces._MapInterface.setup(proxyBinaryMessenger, mapInterfaceController)
    FLTMapInterfaces._AnimationManager.setup(proxyBinaryMessenger, animationController)
    annotationController.setup(proxyBinaryMessenger)
    FLTSettings.LocationComponentSettingsInterface.setup(proxyBinaryMessenger, locationComponentController)
    FLTSettings.LogoSettingsInterface.setup(proxyBinaryMessenger, logoController)
    FLTSettings.GesturesSettingsInterface.setup(proxyBinaryMessenger, gestureController)
    FLTSettings.AttributionSettingsInterface.setup(proxyBinaryMessenger, attributionController)
    FLTSettings.ScaleBarSettingsInterface.setup(proxyBinaryMessenger, scaleBarController)
    FLTSettings.CompassSettingsInterface.setup(proxyBinaryMessenger, compassController)

    methodChannel = MethodChannel(proxyBinaryMessenger, "plugins.flutter.io")
    methodChannel.setMethodCallHandler(this)

    // TODO: check if state-triggered subscription change does not lead to multiple subscriptions/not unsubscribing when listener becomes null
    for (event in eventTypes) {
      subscribeToEvent(FLTMapInterfaces._MapEvent.values()[event])
    }
  }

  private fun subscribeToEvent(event: FLTMapInterfaces._MapEvent) {
    // check deserialization of these events, as they are separate structs declared in GL Native and Flutter plugin
    when (event) {
      FLTMapInterfaces._MapEvent.MAP_LOADED -> mapboxMap.subscribeMapLoaded {
        methodChannel.invokeMethod(event.methodName, Gson().toJson(it))
      }
      FLTMapInterfaces._MapEvent.MAP_LOADING_ERROR -> mapboxMap.subscribeMapLoadingError {
        methodChannel.invokeMethod(event.methodName, Gson().toJson(it))
      }
      FLTMapInterfaces._MapEvent.STYLE_LOADED -> mapboxMap.subscribeStyleLoaded {
        methodChannel.invokeMethod(event.methodName, Gson().toJson(it))
      }
      FLTMapInterfaces._MapEvent.STYLE_DATA_LOADED -> mapboxMap.subscribeStyleDataLoaded {
        methodChannel.invokeMethod(event.methodName, Gson().toJson(it))
      }
      FLTMapInterfaces._MapEvent.CAMERA_CHANGED -> mapboxMap.subscribeCameraChanged {
        methodChannel.invokeMethod(event.methodName, Gson().toJson(it))
      }
      FLTMapInterfaces._MapEvent.MAP_IDLE -> mapboxMap.subscribeMapIdle {
        methodChannel.invokeMethod(event.methodName, Gson().toJson(it))
      }
      FLTMapInterfaces._MapEvent.SOURCE_ADDED -> mapboxMap.subscribeSourceAdded {
        methodChannel.invokeMethod(event.methodName, Gson().toJson(it))
      }
      FLTMapInterfaces._MapEvent.SOURCE_REMOVED -> mapboxMap.subscribeSourceRemoved {
        methodChannel.invokeMethod(event.methodName, Gson().toJson(it))
      }
      FLTMapInterfaces._MapEvent.SOURCE_DATA_LOADED -> mapboxMap.subscribeSourceDataLoaded {
        methodChannel.invokeMethod(event.methodName, Gson().toJson(it))
      }
      FLTMapInterfaces._MapEvent.STYLE_IMAGE_MISSING -> mapboxMap.subscribeStyleImageMissing {
        methodChannel.invokeMethod(event.methodName, Gson().toJson(it))
      }
      FLTMapInterfaces._MapEvent.STYLE_IMAGE_REMOVE_UNUSED -> mapboxMap.subscribeStyleImageRemoveUnused {
        methodChannel.invokeMethod(event.methodName, Gson().toJson(it))
      }
      FLTMapInterfaces._MapEvent.RENDER_FRAME_STARTED -> mapboxMap.subscribeRenderFrameStarted {
        methodChannel.invokeMethod(event.methodName, Gson().toJson(it))
      }
      FLTMapInterfaces._MapEvent.RENDER_FRAME_FINISHED -> mapboxMap.subscribeRenderFrameFinished {
        methodChannel.invokeMethod(event.methodName, Gson().toJson(it))
      }
      FLTMapInterfaces._MapEvent.RESOURCE_REQUEST -> mapboxMap.subscribeResourceRequest {
        methodChannel.invokeMethod(event.methodName, Gson().toJson(it))
      }
    }
  }

  override fun getView(): View {
    return mapView
  }

  override fun dispose() {
    lifecycleProvider.getLifecycle()?.removeObserver(this)
    mapView.onStop()
    mapView.onDestroy()
    methodChannel.setMethodCallHandler(null)
    FLTMapInterfaces.StyleManager.setup(proxyBinaryMessenger, null)
    FLTMapInterfaces._CameraManager.setup(proxyBinaryMessenger, null)
    FLTMapInterfaces.Projection.setup(proxyBinaryMessenger, null)
    FLTMapInterfaces._MapInterface.setup(proxyBinaryMessenger, null)
    FLTMapInterfaces._AnimationManager.setup(proxyBinaryMessenger, null)
    annotationController.dispose(proxyBinaryMessenger)
    FLTSettings.LocationComponentSettingsInterface.setup(proxyBinaryMessenger, null)
    FLTSettings.LogoSettingsInterface.setup(proxyBinaryMessenger, null)
    FLTSettings.GesturesSettingsInterface.setup(proxyBinaryMessenger, null)
    FLTSettings.CompassSettingsInterface.setup(proxyBinaryMessenger, null)
    FLTSettings.ScaleBarSettingsInterface.setup(proxyBinaryMessenger, null)
    FLTSettings.AttributionSettingsInterface.setup(proxyBinaryMessenger, null)
  }

  override fun onStart(owner: LifecycleOwner) {
    super.onStart(owner)
    mapView.onStart()
  }

  override fun onStop(owner: LifecycleOwner) {
    super.onStop(owner)
    mapView.onStop()
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "annotation#create_manager" -> {
        annotationController.handleCreateManager(call, result)
      }
      "annotation#remove_manager" -> {
        annotationController.handleRemoveManager(call, result)
      }
      "gesture#add_listeners" -> {
        gestureController.addListeners(proxyBinaryMessenger)
        result.success(null)
      }
      "gesture#remove_listeners" -> {
        gestureController.removeListeners()
        result.success(null)
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  private fun changeUserAgent(version: String) {
    HttpServiceFactory.setHttpServiceInterceptor(
      object : HttpServiceInterceptorInterface {
        override fun onRequest(
          request: HttpRequest,
          continuation: HttpServiceInterceptorRequestContinuation
        ) {
          request.headers["user-agent"] = "${request.headers["user-agent"]} Flutter Plugin/$version"
          continuation.run(HttpRequestOrResponse(request))
        }

        override fun onResponse(
          response: HttpResponse,
          continuation: HttpServiceInterceptorResponseContinuation
        ) {
          continuation.run(response)
        }
      }
    )
  }
}

private val FLTMapInterfaces._MapEvent.methodName: String
  get() = "event#$ordinal"

File: ios/Classes/MapboxMapController.swift
import Flutter
@_spi(Experimental) import MapboxMaps
import UIKit

struct SuffixBinaryMessenger {
    let messenger: FlutterBinaryMessenger
    let suffix: String
}

final class MapboxMapController: NSObject, FlutterPlatformView {
    private let mapView: MapView
    private let mapboxMap: MapboxMap
    private let channel: FlutterMethodChannel
    private let annotationController: AnnotationController?
    private let gesturesController: GesturesController?
    private let eventHandler: MapboxEventHandler
    private let binaryMessenger: SuffixBinaryMessenger

    func view() -> UIView {
        return mapView
    }

    init(
        withFrame frame: CGRect,
        mapInitOptions: MapInitOptions,
        channelSuffix: Int,
        registrar: FlutterPluginRegistrar,
        pluginVersion: String,
        eventTypes: [Int]
    ) {
        binaryMessenger = SuffixBinaryMessenger(messenger: registrar.messenger(), suffix: String(channelSuffix))
        _ = SettingsServiceFactory.getInstanceFor(.nonPersistent)
            .set(key: "com.mapbox.common.telemetry.internal.custom_user_agent_fragment", value: "FlutterPlugin/\(pluginVersion)")

        mapView = MapView(frame: frame, mapInitOptions: mapInitOptions)
        mapboxMap = mapView.mapboxMap

        channel = FlutterMethodChannel(
            name: "plugins.flutter.io",
            binaryMessenger: binaryMessenger.messenger
        )
        self.eventHandler = MapboxEventHandler(
            eventProvider: mapboxMap,
            binaryMessenger: binaryMessenger.messenger,
            eventTypes: eventTypes
        )

        let styleController = StyleController(styleManager: mapboxMap)
        StyleManagerSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: styleController, messageChannelSuffix: binaryMessenger.suffix)

        let cameraController = CameraController(withMapboxMap: mapboxMap)
        _CameraManagerSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: cameraController, messageChannelSuffix: binaryMessenger.suffix)

        let mapInterfaceController = MapInterfaceController(withMapboxMap: mapboxMap, mapView: mapView)
        _MapInterfaceSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: mapInterfaceController, messageChannelSuffix: binaryMessenger.suffix)

        let mapProjectionController = MapProjectionController()
        ProjectionSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: mapProjectionController, messageChannelSuffix: binaryMessenger.suffix)

        let animationController = AnimationController(withMapView: mapView)
        _AnimationManagerSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: animationController, messageChannelSuffix: binaryMessenger.suffix)

        let locationController = LocationController(withMapView: mapView)
        _LocationComponentSettingsInterfaceSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: locationController, messageChannelSuffix: binaryMessenger.suffix)

        gesturesController = GesturesController(withMapView: mapView)
        GesturesSettingsInterfaceSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: gesturesController, messageChannelSuffix: binaryMessenger.suffix)

        let logoController = LogoController(withMapView: mapView)
        LogoSettingsInterfaceSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: logoController, messageChannelSuffix: binaryMessenger.suffix)

        let attributionController = AttributionController(withMapView: mapView)
        AttributionSettingsInterfaceSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: attributionController, messageChannelSuffix: binaryMessenger.suffix)

        let compassController = CompassController(withMapView: mapView)
        CompassSettingsInterfaceSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: compassController, messageChannelSuffix: binaryMessenger.suffix)

        let scaleBarController = ScaleBarController(withMapView: mapView)
        ScaleBarSettingsInterfaceSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: scaleBarController, messageChannelSuffix: binaryMessenger.suffix)

        annotationController = AnnotationController(withMapView: mapView)
        annotationController!.setup(binaryMessenger: binaryMessenger)

        super.init()

        channel.setMethodCallHandler { [weak self] in self?.onMethodCall(methodCall: $0, result: $1) }
    }

    func onMethodCall(methodCall: FlutterMethodCall, result: @escaping FlutterResult) {
        switch methodCall.method {
        case "annotation#create_manager":
            annotationController!.handleCreateManager(methodCall: methodCall, result: result)
        case "annotation#remove_manager":
            annotationController!.handleRemoveManager(methodCall: methodCall, result: result)
        case "gesture#add_listeners":
            gesturesController!.addListeners(messenger: binaryMessenger)
            result(nil)
        case "gesture#remove_listeners":
            gesturesController!.removeListeners()
            result(nil)
        case "platform#releaseMethodChannels":
            releaseMethodChannels()
            result(nil)
        case "map#snapshot":
            do {
                let snapshot = try mapView.snapshot()
                result(snapshot.pngData())
            } catch {
                result(FlutterError(code: "2342345", message: error.localizedDescription, details: nil))
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func releaseMethodChannels() {
        channel.setMethodCallHandler(nil)
        StyleManagerSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: nil)
        _CameraManagerSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: nil)
        _MapInterfaceSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: nil)
        ProjectionSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: nil)
        _AnimationManagerSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: nil)
        _LocationComponentSettingsInterfaceSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: nil)
        GesturesSettingsInterfaceSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: nil)
        LogoSettingsInterfaceSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: nil)
        AttributionSettingsInterfaceSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: nil)
        CompassSettingsInterfaceSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: nil)
        ScaleBarSettingsInterfaceSetup.setUp(binaryMessenger: binaryMessenger.messenger, api: nil)
        annotationController?.tearDown(messenger: binaryMessenger.messenger)
    }
}
